README.md
# PHP smoothed peak detection algorithm (ZScore)

[![Latest Version on Packagist](https://img.shields.io/packagist/v/gtjamesa/php-zscore.svg?style=flat-square)](https://packagist.org/packages/gtjamesa/php-zscore)
[![Total Downloads](https://img.shields.io/packagist/dt/gtjamesa/php-zscore.svg?style=flat-square)](https://packagist.org/packages/gtjamesa/php-zscore)

This package is the implementation of the "[Robust peak detection algorithm (using z-scores)](https://stackoverflow.com/a/22640362/6029703)" algorithm by [Jean-Paul](https://www.linkedin.com/in/jpgvb).

## Installation

You can install the package via composer:

```bash
composer require gtjamesa/php-zscore
```

## Usage

Parameters to configure the algorithm can be set using a fluent interface, or optionally be specified as the second parameter when creating the object.

``` php
// Create object using fluent interface
$zScore = (new ZScore())->lag(30)
    ->threshold(5)
    ->influence(0);

// Create object using options as an array
$zScore = new ZScore([
    'lag'       => 30,
    'threshold' => 5,
    'influence' => 0,
]);

// Calculate peak signals for the supplied dataset
// Returns [0, 1, 0, 0, ..., -1, 0, 0, 1]
$results = $zScore->calculate($data);
```

The algorithm will iterate the entire supplied dataset on first calculation. This will be inefficient to run on a real-time stream of incoming data, as it will recalculate every signal for the dataset. 

Additional signals for incoming datapoints can be added to the dataset after it's creation by calling the `add()` method:

```php
$signal = $zScore->add(155); // returns -1, 0 or 1
```

#### Saving/loading

The `ZScore` class can be serialized/unserialized so that the previously calculated data may be saved to disk or cache, to be reloaded at a later time. You may optionally call the `shrink()` method before serialization to ignore the dataset, as it is unnecessary for future signal calculations.

```php
$serialized = serialize($zScore); // 3766 bytes
$serializedSmall = serialize($zScore->shrink()); // 3081 bytes
```

#### Algorithm configuration parameters

***`lag`***: the lag parameter determines how much your data will be smoothed and how adaptive the algorithm is to changes in the long-term average of the data. The more [stationary](https://en.wikipedia.org/wiki/Stationary_process) your data is, the more lags you should include (this should improve the robustness of the algorithm). If your data contains time-varying trends, you should consider how quickly you want the algorithm to adapt to these trends. i.e., if you put `lag` at 10, it takes 10 'periods' before the algorithm's threshold is adjusted to any systematic changes in the long-term average. So choose the `lag` parameter based on the trending behaviour of your data and how adaptive you want the algorithm to be.

***`influence`***: this parameter determines the influence of signals on the algorithm's detection threshold. If put at 0, signals have no influence on the threshold, such that future signals are detected based on a threshold that is calculated with a mean and standard deviation that is not influenced by past signals. Another way to think about this is that if you put the influence at 0, you implicitly assume stationarity (i.e. no matter how many signals there are, the time series always returns to the same average over the long term). If this is not the case, you should put the influence parameter somewhere between 0 and 1, depending on the extent to which signals can systematically influence the time-varying trend of the data. e.g., if signals lead to a [structural break](https://en.wikipedia.org/wiki/Structural_break) of the long-term average of the time series, the influence parameter should be put high (close to 1) so the threshold can adjust to these changes quickly.

***`threshold`***: the threshold parameter is the number of standard deviations from the moving mean above which the algorithm will classify a new datapoint as being a signal. For example, if a new datapoint is 4.0 standard deviations above the moving mean and the threshold parameter is set as 3.5, the algorithm will identify the datapoint as a signal. This parameter should be set based on how many signals you expect. For example, if your data is normally distributed, a threshold (or: z-score) of 3.5 corresponds to a signalling probability of 0.00047 (from [this table](https://imgur.com/a/UJlXNJo)), which implies that you expect a signal once every 2128 datapoints (1/0.00047). The threshold therefore directly influences how sensitive the algorithm is and thereby also how often the algorithm signals. Examine your own data and determine a sensible threshold that makes the algorithm signal when you want it to (some trial-and-error might be needed here to get to a good threshold for your purpose).

### Testing

``` bash
composer test
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

### Security

If you discover any security related issues, please email aus.james@gmail.com instead of using the issue tracker.

## Credits

- [James](https://github.com/gtjamesa)
- [Jean-Paul](https://stackoverflow.com/a/22640362/6029703)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `php-zscore` will be documented in this file

## 1.0.0 - 2020-07-06

- Initial release


tests/StatsTest.php
<?php

namespace JamesAusten\PhpZscore\Tests;

use JamesAusten\PhpZscore\Stats;
use PHPUnit\Framework\TestCase;

class StatsTest extends TestCase
{
    /** @test */
    public function can_calculate_mean(): void
    {
        $mean = Stats::mean([2, 4, 4, 4, 5, 5, 7, 9]);
        $this->assertSame(5, $mean);
    }

    /** @test */
    public function can_calculate_stdDev(): void
    {
        $stdDev = Stats::stdDev([2, 4, 4, 4, 5, 5, 7, 9]);
        $this->assertSame(2.0, $stdDev);
    }

    /** @test */
    public function can_calculate_stdDev_from_sample(): void
    {
        $stdDev = Stats::stdDev([2, 4, 4, 4, 5, 5, 7, 9], true);
        $this->assertSame(2.14, round($stdDev, 2));
    }
}


tests/ZScoreTest.php
<?php

namespace JamesAusten\PhpZscore\Tests;

use JamesAusten\PhpZscore\ZScore;
use PHPUnit\Framework\TestCase;

class ZScoreTest extends TestCase
{
    private array $data = [1, 7, 1.1, 1, 0.9, 1, 1, 1.1, 1, 0.9, 1, 1.1, 1, 1, 0.9, 1, 1, 1.1, 1, 1, 1, 1, 1.1, 0.9, 1,
                           1.1, 1, 1, 0.9, 1, 1.1, 1, 1, 1.1, 1, 0.8, 0.9, 1, 1.2, 0.9, 1, 1, 1.1, 1.2, 1, 1.5, 10, 3,
                           2, 5, 3, 2, 1, 1, 1, 0.9, 1, 1, 3, 2.6, 4, 3, 3.2, 2, 1, 1, 0.8, 4, 4, 2, 2.5, 1, 1, 1];

    private array $expected = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                               0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0,
                               1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0];

    /** @test */
    public function can_calculate_zscores_30l_5t_0i(): void
    {
        $zScore = new ZScore([
            'lag'       => 30,
            'threshold' => 5,
            'influence' => 0,
        ]);

        $result = $zScore->calculate($this->data);

//        $this->printDataTable($result);

        $this->assertSame($this->expected, $result);
    }

    /** @test */
    public function can_set_options_with_fluent_interface(): void
    {
        $zScore = new ZScore();

        $result = $zScore->lag(30)
            ->threshold(5)
            ->influence(0)
            ->calculate($this->data);

        $this->assertSame($this->expected, $result);
    }

    /** @test */
    public function can_add_to_previous_results(): void
    {
        // Supply first 64 points
        $zScore = (new ZScore())->lag(30)->threshold(5)->influence(0);

        // Gather the final 4 points
        $add = array_slice($this->data, -10);

        // Run initial calculation
        $zScore->calculate(array_slice($this->data, 0, -10));

        // Assert that adding the final 10 points provides the same expected signals
        $this->assertSame(0, $zScore->add($add[0]));
        $this->assertSame(0, $zScore->add($add[1]));
        $this->assertSame(0, $zScore->add($add[2]));
        $this->assertSame(1, $zScore->add($add[3]));
        $this->assertSame(1, $zScore->add($add[4]));
        $this->assertSame(1, $zScore->add($add[5]));
        $this->assertSame(1, $zScore->add($add[6]));
        $this->assertSame(0, $zScore->add($add[7]));
        $this->assertSame(0, $zScore->add($add[8]));
        $this->assertSame(0, $zScore->add($add[9]));
    }

    /** @test */
    public function can_reserialize_and_continue_functioning(): void
    {
        // Supply first 64 points
        $zScore = (new ZScore())->lag(30)->threshold(5)->influence(0);

        // Gather the final 4 points
        $add = array_slice($this->data, -10);

        // Run initial calculation
        $zScore->calculate(array_slice($this->data, 0, -10));

        // Assert that adding the final 10 points provides the same expected signals
        $this->assertSame(0, $zScore->add($add[0]));
        $this->assertSame(0, $zScore->add($add[1]));
        $this->assertSame(0, $zScore->add($add[2]));
        $this->assertSame(1, $zScore->add($add[3]));

        $serialized = serialize($zScore);
        unset($zScore); // Clear object from memory
        $zScore = unserialize($serialized);

        $this->assertSame(1, $zScore->add($add[4]));
        $this->assertSame(1, $zScore->add($add[5]));
        $this->assertSame(1, $zScore->add($add[6]));
        $this->assertSame(0, $zScore->add($add[7]));
        $this->assertSame(0, $zScore->add($add[8]));
        $this->assertSame(0, $zScore->add($add[9]));

        $this->assertSame(4003, strlen($serialized));
    }

    /** @test */
    public function can_reserialize_shrinked_object(): void
    {
        // Supply first 64 points
        $zScore = (new ZScore())->lag(30)->threshold(5)->influence(0);

        $add = array_slice($this->data, -10);

        $zScore->calculate(array_slice($this->data, 0, -10));

        $serialized = serialize($zScore->shrink());
        $zScore = unserialize($serialized);

        $this->assertSame(0, $zScore->add($add[0]));

        $this->assertSame(3081, strlen($serialized));
    }

    private function printDataTable($result): void
    {
        echo "i\tData\tSignal\n----------------------\n";

        for ($i = 0, $iMax = count($result); $i < $iMax; $i++) {
            echo $i . "\t" . $this->data[$i] . "\t" . $result[$i] . "\n";
        }

        echo "\n\n\n";
    }
}


src/Stats.php
<?php

namespace JamesAusten\PhpZscore;

/**
 * Function implemenations for Stats packages
 *
 * @package JamesAusten\PhpZscore
 */
class Stats
{
    /**
     * Calculate the mean for a set of data
     *
     * @param array $data
     *
     * @return float|int
     */
    public static function mean(array $data)
    {
        return array_sum($data) / count($data);
    }

    /**
     * Calculate standard deviation for a set of data
     *
     * @param array $data
     * @param bool  $sample
     *
     * @return float|int
     */
    public static function stdDev(array $data, bool $sample = false)
    {
        $len = $sample ? count($data) - 1 : count($data);
        $mean = self::mean($data);
        $variance = 0.0;

        foreach ($data as $datum) {
            $variance += self::stdSquare($datum, $mean);
        }

        $variance /= $len;

        return sqrt($variance);
    }

    /**
     * Calculate square of value - mean
     *
     * @param $x
     * @param $mean
     *
     * @return float|int
     */
    public static function stdSquare($x, $mean)
    {
        return ($x - $mean) ** 2;
    }
}


src/ZScore.php
<?php

namespace JamesAusten\PhpZscore;

use Serializable;

/**
 * Implementation of Jean-Paul's ZSCORE algorithm
 *
 * @see     https://stackoverflow.com/a/22640362/6029703
 * @package JamesAusten\PhpZscore
 */
class ZScore implements Serializable
{
    private array $data;

    private int $len;
    private int $lag;
    private float $threshold;
    private float $influence;

    private array $signals;
    private array $avgFilter;
    private array $stdFilter;
    private array $filteredY;

    private bool $shrink = false;

    private array $defaultOptions = [
        'lag'       => 5,
        'threshold' => 3.5,
        'influence' => 0.5,
    ];

    public function __construct(array $options = [])
    {
        $this->mergeOptions($options);
    }

    /**
     * Calculate all signals for a given dataset
     *
     * @param array $data
     *
     * @return array
     */
    public function calculate(array $data): array
    {
        $this->data = $data;

        $this->len = count($this->data);
        $lagData = array_slice($this->data, 0, $this->lag);

        $this->signals = [];
        $this->avgFilter = [];
        $this->stdFilter = [];
        $this->filteredY = [];

        $this->avgFilter[$this->lag - 1] = Stats::mean($lagData);
        $this->stdFilter[$this->lag - 1] = Stats::stdDev($lagData, true);

        for ($i = 0; $i < $this->len; $i++) {
            $this->filteredY[$i] = $this->data[$i];
            $this->signals[$i] = 0;
        }

        for ($i = $this->lag; $i < $this->len; $i++) {
            $this->calcSignal($this->data[$i], $i);
        }

        return $this->signals;
    }

    /**
     * Calculate signal
     *
     * @param $point
     * @param $i
     *
     * @return int
     */
    private function calcSignal($point, $i): int
    {
        if (abs($point - $this->avgFilter[$i - 1]) > $this->threshold * $this->stdFilter[$i - 1]) {
            if ($point > $this->avgFilter[$i - 1]) {
                $this->signals[$i] = 1;
            } else {
                $this->signals[$i] = -1;
            }

            $this->filteredY[$i] = $this->influence * $point + (1 - $this->influence) * $this->filteredY[$i - 1];
        } else {
            $this->signals[$i] = 0;
            $this->filteredY[$i] = $point;
        }

        $lagData = array_slice($this->filteredY, $i - $this->lag, $this->lag);

        $this->avgFilter[$i] = Stats::mean($lagData);
        $this->stdFilter[$i] = Stats::stdDev($lagData, true);

        return $this->signals[$i];
    }

    /**
     * Calculate and return signal for incoming point
     *
     * @param $point
     *
     * @return int
     */
    public function add($point): int
    {
        $this->data[] = $point;
        return $this->calcSignal($point, $this->len++);
    }

    private function mergeOptions(array $options): void
    {
        $options = array_merge($this->defaultOptions, $options);

        $this->lag = (int)$options['lag'];
        $this->threshold = (float)$options['threshold'];
        $this->influence = (float)$options['influence'];
    }

    /**
     * Serialize object, optionally ignoring original supplied data as it is redundant for future calculations
     * To reduce the size of the serialized, call the `shrink` method
     *      serialize($zScore->shrink())
     *
     * @return string
     */
    public function serialize()
    {
        $data = [
            'len'       => $this->len,
            'lag'       => $this->lag,
            'threshold' => $this->threshold,
            'influence' => $this->influence,
            'signals'   => $this->signals,
            'avgFilter' => $this->avgFilter,
            'stdFilter' => $this->stdFilter,
            'filteredY' => $this->filteredY,
        ];

        if (!$this->shrink) {
            $data['data'] = $this->data;
        }

        return serialize($data);
    }

    /** @noinspection UnserializeExploitsInspection */
    public function unserialize($serialized)
    {
        $data = unserialize($serialized);

        $this->data = $data['data'] ?? [];
        $this->len = $data['len'];
        $this->lag = $data['lag'];
        $this->threshold = $data['threshold'];
        $this->influence = $data['influence'];
        $this->signals = $data['signals'];
        $this->avgFilter = $data['avgFilter'];
        $this->stdFilter = $data['stdFilter'];
        $this->filteredY = $data['filteredY'];
    }

    /**
     * Disables serialising the original dataset
     *
     * @return \JamesAusten\PhpZscore\ZScore
     */
    public function shrink(): ZScore
    {
        $this->shrink = true;
        return $this;
    }

    /**
     * @param int $lag
     *
     * @return \JamesAusten\PhpZscore\ZScore
     */
    public function lag(int $lag): ZScore
    {
        $this->lag = $lag;
        return $this;
    }

    /**
     * @param float $threshold
     *
     * @return \JamesAusten\PhpZscore\ZScore
     */
    public function threshold(float $threshold): ZScore
    {
        $this->threshold = $threshold;
        return $this;
    }

    /**
     * @param float $influence
     *
     * @return \JamesAusten\PhpZscore\ZScore
     */
    public function influence(float $influence): ZScore
    {
        $this->influence = $influence;
        return $this;
    }
}
